api/lib/opentelemetry/trace/span_id.rb

```diff
@@ -0,0 +1,50 @@
+# frozen_string_literal: true
+
+# Copyright 2019 OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module Trace
+    # A class that represents a trace identifier. A valid span identifier is an
+    # 8-byte array with at least one non-zero byte.
+    class SpanId
+      INVALID_ID = 0
+      private_constant :INVALID_ID
+
+      MAX_64BIT_UNSIGNED_INT = (1 << 64) - 1
+      private_constant :MAX_64BIT_UNSIGNED_INT
+
+      def self.generate
+        id = rand(MAX_64BIT_UNSIGNED_INT)
+        return generate if id == INVALID_ID
```

Comment:
No need for this to be recursive. You can use a range with `rand` to avoid generating an invalid id.